### PR TITLE
Replace rand() for serial numbers of new certificates with openssl's RAND_bytes.

### PR DESCRIPTION
**Pull request by arsing**
Also change other functions that accept serial numbers to use `long` instead of
`int`. An openssl `ASN1_INTEGER` can be constructed from a `long`, and having
more bits to randomize helps with reducing collisions. The true space of
`ASN1_INTEGER` is much larger but would require going through `BIGNUM`,
which would've made this change more invasive due to needing to manage
the parent `BN_CONTEXT`.

### CODE REVIEW

File: edgelet/hsm-sys/azure-iot-hsm-c/src/edge_pki_openssl.c [hunk 1]
```diff
@@ -1595,22 +1596,34 @@ int generate_pki_cert_and_key_with_props
...
     const char* cert_file_name,
     const char* issuer_key_file,
     const char* issuer_certificate_file
 )
 {
-    return generate_pki_cert_and_key_helper(cert_props_handle,
+    int result;
+
+    long serial_number;
+    if (generate_rand_buffer((unsigned char*)&serial_number, sizeof(serial_number)) != 0)
```
Comment (varunpuranik) [inline edgelet/hsm-sys/azure-iot-hsm-c/src/edge_pki_openssl.c:1609]: Where is generate_rand_buffer implemented?

**Review by varunpuranik — DISMISSED** [review]

**Review by darobs — DISMISSED** [review]

**Review by gordonwang0 — DISMISSED** [review]

File: edgelet/hsm-sys/azure-iot-hsm-c/src/edge_pki_openssl.c [hunk 2]
```diff
@@ -1595,22 +1596,36 @@ int generate_pki_cert_and_key_with_props
...
+    {
+        LOG_ERROR("could not generate serial number");
+        result = __FAILURE__;
+    }
+    else
+    {
+        serial_number &= ((unsigned long)-1) >> 1;
```
Comment (gordonwang0) [inline edgelet/hsm-sys/azure-iot-hsm-c/src/edge_pki_openssl.c:1616]: Making this a call to `labs()` is probably more readable.

Comment (arsing) [reply edgelet/hsm-sys/azure-iot-hsm-c/src/edge_pki_openssl.c:1616]: No, because it doesn't work for `LONG_MIN`.

Comment (gordonwang0) [reply edgelet/hsm-sys/azure-iot-hsm-c/src/edge_pki_openssl.c:1616]: Oh right.

Comment (arsing) [reply edgelet/hsm-sys/azure-iot-hsm-c/src/edge_pki_openssl.c:1620]: Added a comment.

**Review by darobs — APPROVED** [review]

**Review by gordonwang0 — APPROVED** [review]

**Pull request merged**